src/tools/rbd_mirror/LeaderWatcher.h

```diff
@@ -0,0 +1,245 @@
+// -*- mode:C++; tab-width:8; c-basic-offset:2; indent-tabs-mode:t -*-
+// vim: ts=8 sw=2 smarttab
+
+#ifndef CEPH_RBD_MIRROR_LEADER_WATCHER_H
+#define CEPH_RBD_MIRROR_LEADER_WATCHER_H
+
+#include <list>
+#include <memory>
+#include <string>
+
+#include "librbd/ManagedLock.h"
+#include "librbd/managed_lock/Types.h"
+#include "librbd/Watcher.h"
+#include "MirrorStatusWatcher.h"
+#include "tools/rbd_mirror/leader_watcher/Types.h"
+
+namespace librbd { class ImageCtx; }
+
+namespace rbd {
+namespace mirror {
+
+struct Threads;
+
+template <typename ImageCtxT = librbd::ImageCtx>
+class LeaderWatcher : protected librbd::Watcher {
+public:
+  struct Listener {
+    virtual ~Listener() {
+    }
+
+    virtual void post_acquire_handler(Context *on_finish) = 0;
+    virtual void pre_release_handler(Context *on_finish) = 0;
+  };
+
+  LeaderWatcher(Threads *threads, librados::IoCtx &io_ctx);
+
+  void add_listener(Listener *listener);
```

Comment:
Is there a reason to need the add/remove listener methods and not just pass in the listener in the constructor? Is there a one-to-many relationship somewhere?